09-schema/03-rules.md

```diff
@@ -274,6 +274,129 @@ TypeQLDefine query = TypeQL.define(
 </div>
 
 This rule will make every relation transitive.
+## Optimising Transitive Queries
+
+A common use-case for rules is to infer all transitively reachable concepts from a particular (set of) starting concepts (e.g. finding all teams a particular user is a member of recursively, or all nodes reachable from a given node in a graph, etc.) This section describes how to write efficient transitivity rules when it is known in advance which one of the role-players will be specified. This is often the case and we can formulate our rules to answer such queries more efficiently.
+
+The intuitive way of writing a transitive rule is as follows:
+```typeql
+define
+
+node_id sub attribute, value string;
+node sub entity, owns node_id, plays path:from, plays path:to; 
+path sub relation, relates from, relates to;
+
+rule transitive-reachability:
+when {
+    (from: $x, to: $y) isa path;
+    (from: $y, to: $z) isa path;
+} then {
+    (from: $x, to: $z) isa path;
+};
+```
+
+We can interpret this rule as joining two paths together. In a chain `p-q-r-s-t`, querying all nodes reachable from p (`$p isa node, has id "p"; (from: $p, to:$x) isa path;`) would generate the following relations:
+```
+p--q, q--r, r--s, s--t  (already in the database)
+
+p--r, q--s, r--t,       (Inferred)
+p--s, q--t, p--t        (Inferred)                    
+```
+
+Concretely, one `path` relation is generated _for every pair_ of nodes reachable from `p` - a **quadratic** number of relations.
+
+The following section describes an approach which generates only a **linear** number of relations when the `from` role-player is specified. Subsequent sections extend the approach to when the `to` role-player is specified and for symmetric relations where both players play the same role.
+
+### Forward transitivity
+We must first define separate types for the persisted and (inferred) transitive version of the relation.
+For the example above, we use `edge` as the base relation denoting stored facts and `forward-path` as the inferred relation. We then replace the rule with the following two rules: 
+```typeql
+define
+
+node_id sub attribute, value string;
+node sub entity, owns node_id, plays path:from, plays path:to; 
+backward-path sub relation, relates from, relates to;
+
+rule forward-transitivity-base:
+when {
+    (from: $x, to: $y) isa edge;
+} then {
+    (from: $x, to: $y) isa forward-path;
+};
+
+rule forward-transitivity-recursive:
+when {
+    (from: $x, to: $y) isa forward-path;
+    (from: $y, to: $z) isa edge;
+} then {
+    (from: $x, to: $z) isa forward-path;
+};
+```
```

Comment:
after this section and after the next two sections we should show how to trigger the rules in a query correctly